### DISCUSSION TITLE
400 Client Error: Bad Request for url

### DISCUSSION THREAD

**Discussion by BluGeni** [jertel/vuegraf#41] [Q&A] [answered]
I am having trouble getting this up and running. It appears to be logging in and getting my channels. But fails to get usage data. A thought I have is that its because I setup the app with solar and that is where channel "MainsFromGrid" is coming from but not sure what to do to fix it. Any thoughts?

```
vuegraf             | 2021-05-04 14:43:44.632293 | INFO  | Login completed
vuegraf             | 2021-05-04 14:43:44.919825 | INFO  | Discovered new channel: main left (1,2,3)
vuegraf             | 2021-05-04 14:43:45.017391 | INFO  | Discovered new channel: Bath GFI (1)
vuegraf             | 2021-05-04 14:43:45.017975 | INFO  | Discovered new channel: Diner (2)
vuegraf             | 2021-05-04 14:43:45.018410 | INFO  | Discovered new channel: Master bed/smoke/closet (3)
...
vuegraf             | 2021-05-04 14:43:45.271968 | INFO  | Starting as of now 2021-05-04 14:42:36.591662 (last known time is 2021-05-04 14:42:36.591662)
vuegraf             | 2021-05-04 14:43:45.629670 | ERROR | Failed to record new usage data: (<class 'requests.exceptions.HTTPError'>, HTTPError('400 Client Error: Bad Request for url: https://api.emporiaenergy.com/AppAPI?apiMethod=getChartUsage&deviceGid=44463&channel=MainsFromGrid&start=2021-05-04T14:42:36.591662Z&end=2021-05-04T14:43:36.591662Z&scale=1S&energyUnit=KilowattHours'), <traceback object at 0x7f4797285f80>)
```

**Comment by jertel** [comment, marked as answer]
I agree, it appears to be something related to that specific channel. As a temporary troubleshooting step, is it possible for you to remove that channel from the Emporia account or app, and then re-start Vuegraf to see if it gets passed that problem?

**Reply by BluGeni** [reply]
It was something I had specify before setting up any channels. I did some testing directly with PyEmVue with python with this script:

```
from pyemvue import PyEmVue
from pyemvue.enums import Scale
from pyemvue.enums import Unit
import datetime

vue = PyEmVue()
vue.login(username='****@gmail.com', password='*******', token_storage_file='keys.json')
customer = vue.get_customer_details()
print(vars(customer))

device1 = vue.get_devices()[0]
print(device1.device_name)
device1 = vue.populate_device_properties(device1)
print(device1.device_name)


devices = vue.get_devices()
deviceGids = []
for device in devices:
    deviceGids.append(device.device_gid)

channel_usage_list = vue.get_devices_usage(deviceGids, None, scale=Scale.DAY.value, unit=Unit.KWH.value)
for channel in channel_usage_list:
    print(channel.usage, 'kwh')


devices = vue.get_devices()

usage_over_time, start_time = vue.get_chart_usage(devices[0].channels[0], datetime.datetime.utcnow()-datetime.timedelta(days=7), datetime.datetime.utcnow(), scale=Scale.DAY.value, unit=Unit.KWH.value)

print('Usage for the last seven days starting', start_time.isoformat())
for usage in usage_over_time:
    print(usage, 'kwh')
```


Gives me a successful output:

```
{'customer_gid': *****, 'email': '********@gmail.com', 'first_name': '*****', 'lastName': '*****', 'created_at': '2021-05-02T21:57:10.735Z'}
main left
main left
0.35234769728965265 kwh
6.309820293923835 kwh
-5.957472596634183 kwh
0.0 kwh
0.0 kwh
0.3282839635976155 kwh
2.5526743041780255 kwh
0.09177777712663016 kwh
0.5708194161510467 kwh
0.2804038527780109 kwh
6.284194787343343e-06 kwh
0.15915653495748838 kwh
1.903907861964438 kwh
0.013545365980996026 kwh
0.0 kwh
0.0030600500998232095 kwh
0.003459784695174959 kwh
0.408520607238743 kwh
0.0 kwh
0.46485767893791197 kwh
0.0 kwh
1.4813006586498684e-05 kwh
0.2137909454536438 kwh
1.3841769033008151e-05 kwh
2.677083174387614e-05 kwh
-9.989007520469825 kwh
0.408073646895753 kwh
0.22429618876099586 kwh
0.18866302975548638 kwh
0.027598833207554286 kwh
0.10818705919663112 kwh
1.3107314772076078e-05 kwh
0.0 kwh
1.932737647688124 kwh
Usage for the last seven days starting 2021-04-27T05:00:00+00:00
None kwh
None kwh
None kwh
None kwh
None kwh
4.203150147158305 kwh
6.306854064481657 kwh
0.35234769728965265 kwh
```

**Reply by BluGeni** [reply]
I also did not name anything 'MainsFromGrid'

**Reply by jertel** [reply]
That example is only fetching metrics for the first channel in the device. See this code in your above pasted snippet: `devices[0].channels[0]`, where-as Vuegraf fetches usage metrics for all channels in the device. When it hits that MainsFromGrid channel Emporia rejects the request.

**Reply by jmillican61** [reply]
I cannot add much to BlueGeni's post except that I am also getting the same error with MainsFromGrid.  I ran(copy and paste and add proper creds,thank you BlueGeni) the same code as above and got successful output.  Maybe Emporia changed something?  Last successful data pull was yesterday(May, 03 2021) at 09:14 AKDT (17:14 UTC). I had made no changes to running script, simply began to return error.

**Reply by jertel** [reply]
Are you saying Vuegraf used to work, and then stopped working yesterday? If so, did Vuegraf get restarted sometime in the past 24h or so? Also, how long ago did you setup your Emporia devices? 

I just restarted my Vuegraf container and it's still fetching metrics fine. I do not see any channel named "MainsFromGrid" in my log output. So I wonder if this is only occurring with certain, newer devices from Emporia.

**Reply by BluGeni** [reply]
I changed it to this:
```
devices = vue.get_devices()
for device in devices:
    for channel in device.channels:
        print(vars(channel))
```

Output:

```
{'device_gid': 12345, 'name': None, 'channel_num': '1,2,3', 'channel_multiplier': 1.0, 'channel_type_gid': None}
{'device_gid': 12345, 'name': 'Bath GFI', 'channel_num': '1', 'channel_multiplier': 1.0, 'channel_type_gid': 20}
{'device_gid': 12345, 'name': 'Diner', 'channel_num': '2', 'channel_multiplier': 1.0, 'channel_type_gid': 20}
{'device_gid': 12345, 'name': 'Master bed/smoke/closet', 'channel_num': '3', 'channel_multiplier': 1.0, 'channel_type_gid': 20}
{'device_gid': 12345, 'name': 'Nerd Lab', 'channel_num': '4', 'channel_multiplier': 1.0, 'channel_type_gid': 20}
{'device_gid': 12345, 'name': 'Beds', 'channel_num': '5', 'channel_multiplier': 1.0, 'channel_type_gid': 20}
{'device_gid': 12345, 'name': 'LL Outlets', 'channel_num': '6', 'channel_multiplier': 1.0, 'channel_type_gid': 20}
{'device_gid': 12345, 'name': 'Excercise equipment', 'channel_num': '7', 'channel_multiplier': 1.0, 'channel_type_gid': 18}
{'device_gid': 12345, 'name': 'Dryer 2 (red)', 'channel_num': '8', 'channel_multiplier': 1.0, 'channel_type_gid': 4}
{'device_gid': 12345, 'name': 'Kitchen GFI', 'channel_num': '9', 'channel_multiplier': 1.0, 'channel_type_gid': 16}
{'device_gid': 12345, 'name': 'Furnace', 'channel_num': '10', 'channel_multiplier': 1.0, 'channel_type_gid': 11}
{'device_gid': 12345, 'name': 'Microwave', 'channel_num': '11', 'channel_multiplier': 1.0, 'channel_type_gid': 17}
{'device_gid': 12345, 'name': 'Dishwasher and disposal', 'channel_num': '12', 'channel_multiplier': 1.0, 'channel_type_gid': 8}
{'device_gid': 12345, 'name': 'Kitchen GFI', 'channel_num': '13', 'channel_multiplier': 1.0, 'channel_type_gid': 16}
{'device_gid': 12345, 'name': 'LL bath unfinished area', 'channel_num': '14', 'channel_multiplier': 1.0, 'channel_type_gid': 20}
{'device_gid': 12345, 'name': 'Fridge', 'channel_num': '15', 'channel_multiplier': 1.0, 'channel_type_gid': 10}
{'device_gid': 12345, 'name': 'Oven', 'channel_num': '16', 'channel_multiplier': 2.0, 'channel_type_gid': 7}
{'device_gid': 67890, 'name': None, 'channel_num': '1,2,3', 'channel_multiplier': 1.0, 'channel_type_gid': None}
{'device_gid': 67890, 'name': 'Sump', 'channel_num': '1', 'channel_multiplier': 1.0, 'channel_type_gid': 19}
{'device_gid': 67890, 'name': 'Garage GFI', 'channel_num': '2', 'channel_multiplier': 1.0, 'channel_type_gid': 12}
{'device_gid': 67890, 'name': 'Garage Fridge', 'channel_num': '3', 'channel_multiplier': 1.0, 'channel_type_gid': 10}
{'device_gid': 67890, 'name': 'LL Bath GFI', 'channel_num': '4', 'channel_multiplier': 1.0, 'channel_type_gid': 20}
{'device_gid': 67890, 'name': 'Kitchen Lights', 'channel_num': '5', 'channel_multiplier': 1.0, 'channel_type_gid': 20}
{'device_gid': 67890, 'name': 'Dryer', 'channel_num': '6', 'channel_multiplier': 1.0, 'channel_type_gid': 4}
{'device_gid': 67890, 'name': 'A/C', 'channel_num': '7', 'channel_multiplier': 2.0, 'channel_type_gid': 1}
{'device_gid': 67890, 'name': 'PV Solar', 'channel_num': '8', 'channel_multiplier': 2.0, 'channel_type_gid': 13}
{'device_gid': 67890, 'name': 'Weatherproofs LL Outlets', 'channel_num': '9', 'channel_multiplier': 1.0, 'channel_type_gid': 20}
{'device_gid': 67890, 'name': 'Family', 'channel_num': '10', 'channel_multiplier': 1.0, 'channel_type_gid': 20}
{'device_gid': 67890, 'name': 'Dine and Deck Lights', 'channel_num': '11', 'channel_multiplier': 1.0, 'channel_type_gid': 20}
{'device_gid': 67890, 'name': 'Entry/landing/stairs', 'channel_num': '12', 'channel_multiplier': 1.0, 'channel_type_gid': 20}
{'device_gid': 67890, 'name': 'Garage laundry LL lights', 'channel_num': '13', 'channel_multiplier': 1.0, 'channel_type_gid': 20}
{'device_gid': 67890, 'name': 'Bath lights master guest', 'channel_num': '14', 'channel_multiplier': 1.0, 'channel_type_gid': 20}
{'device_gid': 67890, 'name': 'Clothes Washer', 'channel_num': '15', 'channel_multiplier': 1.0, 'channel_type_gid': 5}
{'device_gid': 67890, 'name': 'Dust 2', 'channel_num': '16', 'channel_multiplier': 1.0, 'channel_type_gid': 12}
```

**Reply by jmillican61** [reply]
vuegraf was running perfectly from 2021-04-28 22:30 UTC til it stopped on 2021-05-03 17:14 UTC.  I had restarted a couple times on 2021-05-02 around 02:00 UTC and ran from then till  2021-05-03 17:14 UTC.  I have a Emporia Vue Gen 2 with 16 sensors but have only about half being used so far.  I purchased the device in January of 2021.

**Reply by jertel** [reply]
That's odd. The first post shows a Gid that doesn't appear in your output above.

`&deviceGid=44463`

**Reply by jmillican61** [reply]
Swiping BlueGeni's code again with the update yields:
{'customer_gid': xxxxx, 'email': 'john@xxxxx.xxx', 'first_name': 'John', 'lastName': 'Millican', 'created_at': '2021-01-27T04:03:45.071Z'}
Gorsuch
Gorsuch
10.446826800306374 kwh
10.446826800306374 kwh
0.0 kwh
0.0 kwh
0.22090271999782984 kwh
1.0908083984947206 kwh
7.463182421569824 kwh
1.2114308429993526 kwh
7.160866104527447 kwh
4.158114993659125 kwh
0.00025622046060032316 kwh
0.6448499791916928 kwh
0.9993364562438447 kwh
1.6730004980950575 kwh
{'device_gid': 28440, 'name': None, 'channel_num': '1,2,3', 'channel_multiplier': 1.0, 'channel_type_gid': None}
{'device_gid': 28440, 'name': 'dryer', 'channel_num': '1', 'channel_multiplier': 2.0, 'channel_type_gid': 4}
{'device_gid': 28440, 'name': 'master bath', 'channel_num': '2', 'channel_multiplier': 1.0, 'channel_type_gid': 18}
{'device_gid': 28440, 'name': 'into house solar red', 'channel_num': '3', 'channel_multiplier': 1.0, 'channel_type_gid': 21}
{'device_gid': 28440, 'name': 'into house solar black', 'channel_num': '4', 'channel_multiplier': 1.0, 'channel_type_gid': 21}
{'device_gid': 28440, 'name': 'out to solar red', 'channel_num': '5', 'channel_multiplier': 1.0, 'channel_type_gid': 21}
{'device_gid': 28440, 'name': 'out to solar black', 'channel_num': '6', 'channel_multiplier': 1.0, 'channel_type_gid': 21}
{'device_gid': 28440, 'name': 'Garage Red', 'channel_num': '7', 'channel_multiplier': 1.0, 'channel_type_gid': 21}
{'device_gid': 28440, 'name': 'Garage Black', 'channel_num': '8', 'channel_multiplier': 1.0, 'channel_type_gid': 21}
{'device_gid': 28440, 'name': None, 'channel_num': '9', 'channel_multiplier': 1.0, 'channel_type_gid': None}
{'device_gid': 28440, 'name': None, 'channel_num': '10', 'channel_multiplier': 1.0, 'channel_type_gid': None}
{'device_gid': 28440, 'name': None, 'channel_num': '11', 'channel_multiplier': 1.0, 'channel_type_gid': None}
{'device_gid': 28440, 'name': None, 'channel_num': '12', 'channel_multiplier': 1.0, 'channel_type_gid': None}
{'device_gid': 28440, 'name': None, 'channel_num': '13', 'channel_multiplier': 1.0, 'channel_type_gid': None}
{'device_gid': 28440, 'name': None, 'channel_num': '14', 'channel_multiplier': 1.0, 'channel_type_gid': None}
{'device_gid': 28440, 'name': None, 'channel_num': '15', 'channel_multiplier': 1.0, 'channel_type_gid': None}
{'device_gid': 28440, 'name': None, 'channel_num': '16', 'channel_multiplier': 1.0, 'channel_type_gid': None}
{'device_gid': 29620, 'name': None, 'channel_num': '1,2,3', 'channel_multiplier': 1.0, 'channel_type_gid': 23}
{'device_gid': 42963, 'name': None, 'channel_num': '1,2,3', 'channel_multiplier': 1.0, 'channel_type_gid': 23}
{'device_gid': 43006, 'name': None, 'channel_num': '1,2,3', 'channel_multiplier': 1.0, 'channel_type_gid': 23}
Usage for the last seven days starting 2021-04-27T08:00:00+00:00
17.551346113905907 kwh
14.929532480012575 kwh
24.2266524026129 kwh
21.167340488239923 kwh
18.461780943769746 kwh
16.879317771877712 kwh
19.13967389640172 kwh
10.446826800306374 kwh
jmillican@desk1:~/PyEmVue$ 

Hope this is useful, if I am just distracting, please let me know.   I am at best a guess and check hobby dev.

**Reply by BluGeni** [reply]
> That's odd. The first post shows a Gid that doesn't appear in your output above.
> 
> `&deviceGid=44463`

Sorry I had just changed them for anonymity. They match that number.

**Reply by jertel** [reply]
I just pushed a change to Vuegraf that will print more debug logging when this error occurs, and it will also keep going with the other channels instead of exiting. The new Docker image is ready (`jertel/vuegraf:latest`) if you're using Docker. Let me know what you both find after trying the new version.

**Reply by jmillican61** [reply]
I am running as a host process.  I can run as docker but will take a bit to get it all set up. Will update when done and tested.  Thank you.

**Reply by jertel** [reply]
You don't need to switch to Docker. If you pull the latest into your vuegraf dir you can restart your host process to pick up the new change.

**Reply by jmillican61** [reply]
Awesome.  Thanks.

**Reply by jmillican61** [reply]
2021-05-04 17:23:48.644762 | INFO  | Using InfluxDB version 2
2021-05-04 17:23:53.023266 | INFO  | Login completed
2021-05-04 17:23:53.868064 | INFO  | Discovered new channel: Gorsuch (1,2,3)
2021-05-04 17:23:54.304101 | INFO  | Discovered new channel: dryer (1)
2021-05-04 17:23:54.304189 | INFO  | Discovered new channel: master bath (2)
2021-05-04 17:23:54.304245 | INFO  | Discovered new channel: into house solar red (3)
2021-05-04 17:23:54.304302 | INFO  | Discovered new channel: into house solar black (4)
2021-05-04 17:23:54.304358 | INFO  | Discovered new channel: out to solar red (5)
2021-05-04 17:23:54.304413 | INFO  | Discovered new channel: out to solar black (6)
2021-05-04 17:23:54.304467 | INFO  | Discovered new channel: Garage Red (7)
2021-05-04 17:23:54.304521 | INFO  | Discovered new channel: Garage Black (8)
2021-05-04 17:23:54.304578 | INFO  | Discovered new channel: None (9)
2021-05-04 17:23:54.304638 | INFO  | Discovered new channel: None (10)
2021-05-04 17:23:54.304696 | INFO  | Discovered new channel: None (11)
2021-05-04 17:23:54.304760 | INFO  | Discovered new channel: None (12)
2021-05-04 17:23:54.304823 | INFO  | Discovered new channel: None (13)
2021-05-04 17:23:54.304861 | INFO  | Discovered new channel: None (14)
2021-05-04 17:23:54.304894 | INFO  | Discovered new channel: None (15)
2021-05-04 17:23:54.304926 | INFO  | Discovered new channel: None (16)
2021-05-04 17:23:54.581384 | INFO  | Discovered new channel: Refrigerator (1,2,3)
2021-05-04 17:23:54.955757 | INFO  | Discovered new channel: John desktop (1,2,3)
2021-05-04 17:23:55.293839 | INFO  | Discovered new channel: testplug (1,2,3)
2021-05-04 17:23:55.331745 | INFO  | Starting as of now 2021-05-04 17:22:43.645495 (last known time is 2021-05-02 23:59:59.275530)
2021-05-04 17:23:56.377821 | ERROR | Failed to fetch metrics: (<class 'requests.exceptions.HTTPError'>, HTTPError('400 Client Error: Bad Request for url: https://api.emporiaenergy.com/AppAPI?apiMethod=getChartUsage&deviceGid=28440&channel=MainsFromGrid&start=2021-05-04T17:22:43.645495Z&end=2021-05-04T17:23:43.645495Z&scale=1S&energyUnit=KilowattHours'), <traceback object at 0x7f739c0e5048>)
2021-05-04 17:23:56.377922 | ERROR | Failed on: deviceGid=28440; chanNum=MainsFromGrid; chanName=Gorsuch;
2021-05-04 17:23:56.807662 | ERROR | Failed to fetch metrics: (<class 'requests.exceptions.HTTPError'>, HTTPError('400 Client Error: Bad Request for url: https://api.emporiaenergy.com/AppAPI?apiMethod=getChartUsage&deviceGid=28440&channel=MainsToGrid&start=2021-05-04T17:22:43.645495Z&end=2021-05-04T17:23:43.645495Z&scale=1S&energyUnit=KilowattHours'), <traceback object at 0x7f739c0e5048>)
2021-05-04 17:23:56.807763 | ERROR | Failed on: deviceGid=28440; chanNum=MainsToGrid; chanName=Gorsuch;
2021-05-04 17:24:00.804953 | INFO  | Submitted datapoints to database; account="Gorsuch"; points=177

**Reply by jertel** [reply]
@jmillican61 You also have a solar setup integrated with Emporia? That is making more sense, since it would explain why it's not happening to me (I don't have a PV system yet).

**Reply by jmillican61** [reply]
I have the 4 circuits for solar set up as a sub panel circuit type.  I use two of the sixteen available sensors to read energy to the Inverter and then two more to read the energy coming back from the Inverter. I know this will skew the main % in the Emporia app somewhat but I Grafana so that it makes sense visually.
Testing with the new pull, I am getting data.  Still with the MainsFromGrid error but there does not appear to be anything missing so far.

**Reply by jertel** [reply]
Are you also getting data points for channels 3 through 8? 

And are the bottom three channels Smart Plugs?

**Reply by jmillican61** [reply]
Yes, I am getting data for all connected sensors and the last three are smart plugs.  The solar stuff is circuits 3,4,5,and 6.  I have an electric clothes dryer as circuit 1 with a 2.0 multiplier and circuit 2 is a bathroom, circuits 7 and 8 are feeds to my Garage. the garage is not seen on the mains as they are split off at the power meter, separate from the Main to the house.   No dedicated Solar monitor.

**Reply by jertel** [reply]
Ok thanks. I'll leave it as-is for now since it's able to fetch the data at this point. It's possible that this is a bug with a new Emporia change so I don't want to dive much deeper yet. If it continues we can check with @magico13 to see if he's got an idea on what's going on. He wrote the library that parses the responses from Emporia.

**Reply by jmillican61** [reply]
Thank you very much.

**Reply by BluGeni** [reply]
This is working with the update you did. Thank you!

**Reply by magico13** [reply]
Just an FYI I pushed an update to PyEmVue (0.12.3) so if you end up passing MainsToGrid or MainsFromGrid into `get_chart_usage` it'll just return empty since those virtual channels don't work along that endpoint. That should hopefully avert this issue, we're having similar issues in the home assistant plugin with these virtual channels. They only exist for users with solar installed apparently and just showed up in the last week or so.
